### PR TITLE
Few modifications for better training with ML-Agents

### PR DESCRIPTION
**Pull request by vincentpierre**
Made a new model and retrained with a different observation and action space.

The new observation space includes the position and orientation of each joints
The new action space is now 7 discrete actions (one per joint) each with 3 different possibilities (Clockwise, Nothing, CounterClockwise)
Increased the Decision period to 20
Used the following configuration for training :
default:
    trainer: sac
    batch_size: 128
    buffer_size: 50000
    buffer_init_steps: 0
    hidden_units: 128
    init_entcoef: 1.0
    learning_rate: 3.0e-4
    learning_rate_schedule: constant
    max_steps: 5.0e5
    memory_size: 128
    normalize: false
    num_update: 1
    train_interval: 1
    num_layers: 2
    time_horizon: 64
    sequence_length: 64
    summary_freq: 10000
    tau: 0.005
    use_recurrent: false
    vis_encode_type: simple
    reward_signals:
        extrinsic:
            strength: 1.0
            gamma: 0.99

TouchCube:
    summary_freq: 1000
    max_steps: 5.0e6
    reward_signals:
        extrinsic:
            strength: 1.0
            gamma: 0.95

Fixed a bug in ActionIndexToRotationDirection (wrong sign in equation)

### CODE REVIEW

File: ArmRobot/Assets/Scripts/RobotAgent.cs [hunk 1]
```diff
@@ -40,46 +40,66 @@ public override void CollectObservations(VectorSensor sensor)
...
-        // end episode if we touched the cube
-        if (touchDetector.hasTouchedTarget)
+        for (int jointIndex = 0; jointIndex < vectorAction.Length; jointIndex ++)
         {
-            EndEpisode();
+            RotationDirection rotationDirection = ActionIndexToRotationDirection((int) vectorAction[jointIndex]);
+            robotController.RotateJoint(jointIndex, rotationDirection, false);
         }
 
+            // end episode if we touched the cube
```
Comment (mmattar) [inline ArmRobot/Assets/Scripts/RobotAgent.cs:74]: this block seems to not be aligned correctly

**Comment by mmattar** [comment]
@vincentpierre - two quick things: 
1. can you update the config file with the new params (assuming you've changed some of them)
2. Can you update the Tensorboard reward image in the README file to reflect the new setting.

**Comment by vincentpierre** [comment]
@mmattar  Done

**Comment by vincentpierre** [comment]
> the weird "glitch" issue went away with this update

That should not have been the case...

**Comment by mmattar** [comment]
> > the weird "glitch" issue went away with this update
> 
> That should not have been the case...

Maybe it was just the random seed then - I presume it's different with different training sessions?

File: ArmRobot/Assets/Scripts/RobotAgent.cs [hunk 2]
```diff
@@ -40,46 +40,66 @@ public override void CollectObservations(VectorSensor sensor)
         float[] rotations = robotController.GetCurrentJointRotations();
         foreach (float rotation in rotations)
         {
-            // normalize rotation to [0, 1] range
-            float normalizedRotation = rotation / 360.0f;
+            // normalize rotation to [-1, 1] range
+            float normalizedRotation = (rotation / 360.0f) %  1f;
```
Comment (sarahwolf32) [inline ArmRobot/Assets/Scripts/RobotAgent.cs:44]: Why did we add the modulo 1?

Comment (vincentpierre) [reply ArmRobot/Assets/Scripts/RobotAgent.cs:44]: Because a rotation of x + 360 = x (corresponds to a full rotation)
It has the advantage to compress the state space better at the cost of having the observation space be discontinuous at 359 degrees

File: ArmRobot/Assets/Scripts/RobotAgent.cs [hunk 3]
```diff
@@ -40,46 +40,66 @@ public override void CollectObservations(VectorSensor sensor)
...
             sensor.AddObservation(normalizedRotation);
         }
 
+        foreach (var joint in robotController.joints)
```
Comment (sarahwolf32) [inline ArmRobot/Assets/Scripts/RobotAgent.cs:48]: We don't really need to loops here, since under the covers, the other loop is just going through the joints as well. We could just use this new loop. We could add
```
ArticulationJointController jointController = joint.robotPart.GetComponent<ArticulationJointController>();
float rotation = jointController.CurrentPrimaryAxisRotation();
float normalizedRotation = (rotation / 360.0f) %  1f;
sensor.AddObservation(normalizedRotation);
```
to it, and then delete the loop above.

File: ArmRobot/Assets/Scripts/RobotAgent.cs [hunk 4]
```diff
@@ -40,46 +40,66 @@ public override void CollectObservations(VectorSensor sensor)
...
         }
 
+
         //reward
-        float distanceToCube = Vector3.Distance(endEffector.transform.position, cube.transform.position);
-        SetReward(-distanceToCube * 0.01f);
+        float distanceToCube = Vector3.Distance(endEffector.transform.position, cube.transform.position); // roughly 0.7f
         
+
+        var jointHeight = 0f; // This is to reward the agent for keeping high up // max is roughly 3.0f
```
Comment (sarahwolf32) [inline ArmRobot/Assets/Scripts/RobotAgent.cs:86]: Good idea! :)

**Comment by sarahwolf32** [comment]
> > > the weird "glitch" issue went away with this update
> > 
> > 
> > That should not have been the case...
> 
> Maybe it was just the random seed then - I presume it's different with different training sessions?

What did this glitch look like? The negative distance-to-cube part of the reward can spike downward sometimes if the cube flies off the table and gets very far away. It _should_ at least reset correctly though. I've thought about adding some other mechanism to keep the cube on the table or using a min reward to combat this, but never implemented it.

File: ArmRobot/Assets/Scripts/RobotAgent.cs (hunk 2 above)
Comment (sarahwolf32) [reply ArmRobot/Assets/Scripts/RobotAgent.cs:44]: It is also discontinuous at 0, right in the middle of our [-1, 1] range, right? Using python as a sanity check, I got:
```
>>> 0.001 % 1
0.001
>>> -0.001 % 1
0.999
```
Also, I think we might be losing some information in our [-1, 1] range:
```
>>> -0.5 % 1
0.5
>>> 0.5 % 1
0.5
```

**Comment by sarahwolf32** [comment]
Looks great! Feel free to merge when you feel ready!

File: ArmRobot/Assets/Scripts/RobotAgent.cs (hunk 2 above)
Comment (sarahwolf32) [reply ArmRobot/Assets/Scripts/RobotAgent.cs:44]: What if we make the rotation range be [0, 1], with the discontinuity at 0/1, but then shift the range to map the current 0 position to 0.5, so that at least there is not a discontinuity at the default position?

Comment (sarahwolf32) [reply ArmRobot/Assets/Scripts/RobotAgent.cs:44]: Or - if we want to be more realistic - we could physically limit the rotation of each joint to [-720, 720 ] degrees (this is the limit on the real robot). Commands to continue rotating past the limit simply don't work. Then we have an enforced range we can normalize over without worrying about discontinuities.

Comment (vincentpierre) [reply ArmRobot/Assets/Scripts/RobotAgent.cs:44]: C# and Python behave differently. -0.5f % 1f = -0.5f not 0.5f (I don't know why)

Comment (sarahwolf32) [reply ArmRobot/Assets/Scripts/RobotAgent.cs:44]: Oh wow, I didn't know that. Ok then. :)

File: README.md [hunk 5]
```diff
@@ -8,8 +8,8 @@ This new joint system, powered by [Nvidia's PhysX 4](https://news.developer.nvid
 
 ## Requirements
 
-Unity 2020.1.0b1 or later is needed for the new joint system. This UR3 demo comes
-bundled with [ML-Agents v0.15.1](https://github.com/Unity-Technologies/ml-agents/tree/0.15.1)
+Unity 2020.1.0b1 or later is needed for the new joint system. This UR3 demo works
+with [ML-Agents release_1](https://github.com/Unity-Technologies/ml-agents/tree/release_1)
```
Comment (mmattar) [inline README.md:12]: make it "ML-Agents Release 1" since that is the human-friendly name

File: README.md [hunk 6]
```diff
@@ -41,9 +41,9 @@ in Unity you will see a few errors since the ML-Agents package still needs to be
 
 #### Install the Unity ML-Agents Toolkit
 
-Detailed instructions for installing v0.15.1 of the ML-Agents Toolkit can be found on the
-[ml-agents repo](https://github.com/Unity-Technologies/ml-agents/blob/0.15.1/docs/Installation.md).
-However, do skip the [Install Unity section](https://github.com/Unity-Technologies/ml-agents/blob/0.15.1/docs/Installation.md#install-unity-20184-or-later) as we've already covered the necessary
+Detailed instructions for installing release_1 of the ML-Agents Toolkit can be found on the
```
Comment (mmattar) [inline README.md:44]: release_1 --> Release 1
